python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,201 @@
+from dataclasses import dataclass
+from typing import Dict, Optional, Tuple, cast
+
+from eth_typing import Hash32
+from ethereum.cancun.fork_types import Account, Address
+from ethereum_types.numeric import U256
+from ethereum_types.bytes import Bytes32
+
+from ethereum.cancun.trie import InternalNode, LeafNode, BranchNode, ExtensionNode
+from ethereum.crypto.hash import Hash32, keccak256
+from ethereum_types.bytes import Bytes, Bytes20, Bytes32
+from ethereum_types.numeric import Uint
+from ethereum_rlp import rlp
+from mpt.utils import nibble_path_to_bytes
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+@dataclass
+class AccountNode:
+    nonce: Uint
+    balance: U256
+    code_hash: Hash32
+    storage_root: Hash32
+
+    @staticmethod
+    def from_rlp(bytes: Bytes) -> "AccountNode":
+        decoded = rlp.decode(bytes)
+        return AccountNode(
+            nonce=Uint(int.from_bytes(decoded[0], "big")),
+            balance=U256(int.from_bytes(decoded[1], "big")),
+            storage_root=Hash32(decoded[2]),
+            code_hash=Hash32(decoded[3]),
+        )
+
+    def to_account(self, code: Bytes) -> Account:
+        return Account(
+            nonce=self.nonce,
+            balance=self.balance,
+            code=code,
+        )
+
+@dataclass
+class StateDiff:
+    """
+    Contains all information that is preserved between transactions.
+    """
+
+    _main_trie: Dict[Address, Tuple[Optional[AccountNode], Optional[AccountNode]]]
+    _storage_tries: Dict[Address, Tuple[Dict[Bytes32, U256], Dict[Bytes32, U256]]]
+
+    _nodes: Dict[Hash32, InternalNode]
+    _address_preimages: Dict[Hash32, Address]
+    _storage_key_preimages: Dict[Hash32, Bytes32]
+
+    @classmethod
+    def from_tries(cls, tries: "EthereumTries") -> "StateDiff":
+        # Merge all mappings of hash -> node to have a single DB to fetch nodes and resolve addresses from
+        diff = StateDiff({}, {}, tries.nodes, tries.address_preimages, tries.storage_key_preimages)
+
+        l_root = tries.pre_state_root
+        r_root = tries.post_state_root
+
+        diff._compute_diff(l_root, r_root, Bytes())
+        return diff
+
+    def _compute_diff(self, left: Optional[Hash32], right: Optional[Hash32], path: Bytes):
+        if left == right:
+            return
+
+        l_node = self._nodes.get(left) if left else None
+        r_node = self._nodes.get(right) if right else None
+
+        l_name = l_node.__class__.__name__ if l_node else "None"
+        r_name = r_node.__class__.__name__ if r_node else "None"
+
+        match l_name, r_name:
+            case "None", "None":
+                # No change
+                pass
+
+            case "None", "LeafNode":
+                # new leaf
+                preimage = nibble_path_to_bytes(path + r_node.rest_of_key)
```

Comment:
are you sure this is the entire path?

imagine you are half way thru your trie, your path is 0x123456

if your full path to a leaf is 0x123456abcde
the leaf' rest_of_key will be 0xabcde